File: C++/src/matrix.cpp
#pragma region LICENSE

// MIT License

// Copyright (c) 2024 Rohan Bharatia

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#pragma endregion LICENSE

#pragma once

#ifndef MATRIX_CPP
#define MATRIX_CPP

// std
#include <cassert>
#include <iostream>
#include <iomanip>

#include "include/matrix.hpp"

Matrix Matrix::add(Matrix &other)
{
    assert(cols == other.rows);
    
    Matrix target(other.cols, other.rows);

    for(uint32_t x = 0; x < target.cols; x++)
    {
        for(uint32_t y = 0; y < target.rows; y++)
        {
            float result = 0.0f;

            for(uint32_t k = 0; k < cols; k++)
                result += at(k, y) + other.at(k, y);

            target.at(x, y) = result;
        }
    }

    return target;
}

Matrix Matrix::add(float &scalar);
{
    assert(cols == other.rows);
    
    Matrix target(other.cols, other.rows);

    for(uint32_t x = 0; x < target.cols; x++)
    {
        for(uint32_t y = 0; y < target.rows; y++)
            target.at(x, y) = at(x, y) + scalar;
    }
    
    return target;
}

Matrix Matrix::subtract(Matrix &other)
{
    assert(cols == other.rows);
    
    Matrix target(other.cols, other.rows);

    for(uint32_t x = 0; x < target.cols; x++)
    {
        for(uint32_t y = 0; y < target.rows; y++)
        {
            float result = 0.0f;

            for(uint32_t k = 0; k < cols; k++)
                result += at(k, y) - other.at(k, y);

            target.at(x, y) = result;
        }
    }

    return target;
}

Matrix Matrix::subtract(float &scalar);
{
    assert(cols == other.rows);
    
    Matrix target(other.cols, other.rows);

    for(uint32_t x = 0; x < target.cols; x++)
    {
        for(uint32_t y = 0; y < target.rows; y++)
            target.at(x, y) = at(x, y) - scalar;
    }
    
    return target;
}

Matrix Matrix::multiply(Matrix &other)
{
    assert(cols == other.rows);
    
    Matrix target(other.cols, other.rows);

    for(uint32_t x = 0; x < target.cols; x++)
    {
        for(uint32_t y = 0; y < target.rows; y++)
        {
            float result = 0.0f;

            for(uint32_t k = 0; k < cols; k++)
                result += at(k, y) * other.at(k, y);

            target.at(x, y) = result;
        }
    }
    
    return target;
}

Matrix Matrix::multiply(float &scalar);
{
    assert(cols == other.rows);
    
    Matrix target(other.cols, other.rows);

    for(uint32_t x = 0; x < target.cols; x++)
    {
        for(uint32_t y = 0; y < target.rows; y++)
            target.at(x, y) = at(x, y) * scalar;
    }
    
    return target;
}

Matrix Matrix::divide(Matrix &other)
{
    assert(cols == other.rows);
    
    Matrix target(other.cols, other.rows);

    for(uint32_t x = 0; x < target.cols; x++)
    {
        for(uint32_t y = 0; y < target.rows; y++)
        {
            if(target.at(x, y) != 0)
            {
                float result = 0.0f;

                for(uint32_t k = 0; k < cols; k++)
                    result += at(k, y) / other.at(k, y);

                target.at(x, y) = result;
            }
            else
            {
                std::runtime_error("Unable divide by zero!");
                return EXIT_FAILURE;
            }
        }
    }

    return target;
}

Matrix Matrix::divide(float &scalar);
{
    assert(cols == other.rows);
    
    Matrix target(other.cols, other.rows);

    for(uint32_t x = 0; x < target.cols; x++)
    {
        for(uint32_t y = 0; y < target.rows; y++)
        {
            if(scalar != 0)
                target.at(x, y) = at(x, y) / scalar;
            else
            {
                std::runtime_error("Unable to divide by zero!");
                return EXIT_FAILURE;
            }
        }
    }
    
    return target;
}

Matrix Matrix::flip_signs()
{
    Matrix target(cols, rows);

    for(uint32_t x = 0; x < target.cols; x++)
    {
        for(uint32_t y = 0; y < target.rows; y++)
            target.at(x, y) = -at(x, y);
    }
    
    return target;
}

Matrix Matrix::transpose()
{
    Matrix target(cols, rows);

    for(uint32_t x = 0; x < cols; x++)
    {
        for(uint32_t y = 0; y < rows; y++)
            target.at(x, y) = at(y, x);
    }
    
    return target;
}

template<typename T>
Matrix Matrix::apply_function(std::function<T(const T&)> func)
{
    Matrix target(cols, rows);

    for(uint32_t x = 0; x < cols; x++)
    {
        for(uint32_t y = 0; y < rows; y++)
            target.at(x, y) = func(at(x, y));
    }

    return target;
}

Matrix Matrix::cofactor(uint32_t col, uint32_t row)
{
    Matrix target(cols - 1, rows - 1);
    uint32_t idx = 0;

    for(uint32_t x = 0; x < cols; x++)
    {
        for(uint32_t y = 0; y < rows; y++)
        {
            if(x == col || y == row)
                continue;

            target.values[idx++] = at(x, y);
        }
    }

    return target;
}

template<typename T>
T Matrix::determinant()
{
    assert(rows == cols);
    T target();

    if(rows == 1)
        return values[0];
    else
    {
        uint32_t sign = 1;

        for(uint32_t x = 0; x < cols; x++)
        {
            target += sign * at(x, 0) * cofactor(x, 0).determinant();
            sign *= -1;
        }
    }

    return target;
}

Matrix Matrix::adjoint()
{
    assert(rows == cols);

    Matrix target(cols, rows);
    uint32_t sign = 1;

    for(uint32_t x = 0; x < cols; x++)
    {
        for(uint32_t y = 0; y < rows; y++)
        {
            target.at(x, y) += sign * cofactor(x, y).determinant();
            sign *= -1;
        }
    }
    
    return target.transpose();
}

template<typename T>
Matrix Matrix::inverse()
{
    Matrix adj = adjoint();
    T factor = determinant();

    for(uint32_t x = 0; x < col; x++)
    {
        for(uint32_t x = 0; x < col; x++)
            adj.at(x, y) /= factor;
    }

    return adj;
}

Mat2x2::Mat2x2()
{
    Matrix matrix(2, 2);
    return matrix;
}

Mat3x3::Mat3x3()
{
    Matrix matrix(3, 3);
    return matrix;
}

Mat4x4::Mat4x4()
{
    Matrix matrix(4, 4);
    return matrix;
}

#endif

File: Python/matrix.py
# MIT License

# Copyright (c) 2024 Rohan Bharatia

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import numpy as np
import sys

class Matrix:
    global cols
    global rows
    nonlocal lst = {}
    global values = np.array(lst)

    def __init__(cols, rows):
        cols = cols
        rows = rows
        values = values[:(cols, rows), 0]

    def at(col, row):
        return values[row * cols + col]
    
    def add(other):
        if isinstance(other, Matrix):
            continue
        
        assert cols == other.rows

        target: Matrix(other.cols, other.rows)

        for x in range(target.cols):
            for y in range(target.rows):
                result = 0

                for k in range(cols):
                    result = at(k, y) + other.at(k, y)
                
                target.at(x, y) = result

        return target
    
    def add(scalar)
        if isinstance(scalar, float):
            continue
        
        assert cols == other.rows
        target: Matrix(other.cols, other.rows)

        for x in range(target.cols):
            for y in range(target.rows):
                target.at(x, y) = at(x, y) + scalar
        
        return target

    def subtract(other):
        if isinstance(other, Matrix):
            continue
        
        assert cols == other.rows

        target: Matrix(other.cols, other.rows)

        for x in range(target.cols):
            for y in range(target.rows):
                result = 0

                for k in range(cols):
                    result = at(k, y) - other.at(k, y)
                
                target.at(x, y) = result

        return target

    def subtract(scalar)
        if isinstance(scalar, float):
            continue
        
        assert cols == other.rows
        target: Matrix(other.cols, other.rows)

        for x in range(target.cols):
            for y in range(target.rows):
                target.at(x, y) = at(x, y) - scalar
        
        return target
    
    def multiply(other):
        if isinstance(other, Matrix):
            continue
        
        assert cols == other.rows

        target: Matrix(other.cols, other.rows)

        for x in range(target.cols):
            for y in range(target.rows):
                result = 0

                for k in range(cols):
                    result = at(k, y) * other.at(k, y)
                
                target.at(x, y) = result

        return target
    
    def multiply(scalar)
        if isinstance(scalar, float):
            continue
        
        assert cols == other.rows
        target: Matrix(other.cols, other.rows)

        for x in range(target.cols):
            for y in range(target.rows):
                target.at(x, y) = at(x, y) * scalar
        
        return target

    def divide(other):
        if isinstance(other, Matrix):
            continue
        
        assert cols == other.rows

        target: Matrix(other.cols, other.rows)

        for x in range(target.cols):
            for y in range(target.rows):
                result = 0

                for k in range(cols):
                    if target.at(x, y) != 0:
                        result = at(k, y) / other.at(k, y)
                    else:
                        print("Unable to divide by zero!", file  = sys.stderr)
                
                target.at(x, y) = result

        return target
    
    def divide(scalar)
        if isinstance(scalar, float):
            continue
        
        assert cols == other.rows
        target: Matrix(other.cols, other.rows)

        for x in range(target.cols):
            for y in range(target.rows):
                if target.at(x, y) != 0:
                    target.at(x, y) = at(x, y) / scalar
                else:
                    print("Unable to divide by zero!", file = sys.stderr)
        
        return target
    
    def flip_signs():
        target: Matrix(cols, rows)

        for x in range(target.cols):
            for y in range(target.rows):
                target.at(x, y) = -at(x, y)

        return target
    
    def transpose():
        target: Matrix(cols, rows)

        for x in range(target.cols):
            for y in range(target.rows):
                target.at(x, y) = at(y, x)

        return target
    
    def apply_function(func):
        target: Matrix(cols, rows)

        for x in range(target.cols):
            for y in range(target.rows):
                target.at(x, y) = func(at(x, y))

        return target
    
    def cofactor(col, row):
        target: Matrix(cols, rows)
        idx = 1

        for x in range(target.cols):
            for y in range(target.rows):
                if x == col or y == row:
                    continue

                target.values[idx] = at(x, y)
                idx = idx + 1

        return target
        
    def determinant():
        assertrows == cols

        target = 0

        ifrows == 1:
            return values[0]
        else:
            sign = 1

            for x in range(cols):
                target = target + sign * at(x, 0) * cofactor(x, 0).determinant()
                sign = sign * -1
            
        return target
    
    def adjoint():
        assertrows == cols

        target: Matrix(cols, rows)
        sign = 1

        for x in range(cols):
            for y in range(rows):
                target.at(x, y) = target + sign * at(x, y) * cofactor(x, y).determinant()
                sign = sign * -1

        return target.transpose()
    
    def inverse():
        adj: Matrix = adjoint()
        factor = determinant()

        for x in range(cols):
            for y in range(rows):
                adj.at(x, y) = adj.at(x, y) / factor
        
        return adj

class Mat2x2(Matrix):
    def __init__():
        Matrix matrix(2, 2)
        return matrix

class Mat3x3(Matrix):
    def __init__():
        Matrix matrix(3, 3)
        return matrix

class Mat4x4(Matrix):
    def __init__():
        Matrix matrix(4, 4)
        return matrix